### PR TITLE
JS: add query: js/unconditional-property-override

### PR DESCRIPTION
**Pull request by esben-semmle**
This adds the heap variant of `js/useless-assignment-to-local`. The query is obviously a bit less precise, but I am impressed with some of the results (see `real-world-examples.js` in the diff to see some non-trivial result patterns).

The query is based on a dominance computation, in particular: whether a write to a property is post-dominated by another write to the same property without an interleaving use of the property.

An [evaluation](https://git.semmle.com/esben/dist-compare-reports/tree/js/unconditional-property-override_1541059627636) shows that the query scales fine (notice that the used query suite is small, the time changes compared to a full query suite run are minor).

There is a lot of false positives in browser test code, apparently the externs do not define all relevant DOM setters. This also affects a few pieces of library code:

```js
if ( jQuery.browser.opera && name == "display" ) {
    var save = style.outline;
    style.outline = "0 solid black"; // flagged
    style.outline = save;
}
```

I think we should see if it becomes a problem in practice before we handle additional setters.

### CODE REVIEW

File: javascript/ql/src/Declarations/DeadStoreOfProperty.ql [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import javascript
+import Expressions.DOMProperties
+import DeadStore
+
+/**
+ * Holds if `assign` definitely assigns property `name` of `base`.
+ */
+predicate unambigiousPropWrite(DataFlow::SourceNode base, string name, Assignment assign) {
+  exists(DataFlow::PropWrite lhs |
+    assign.getLhs().flow() = lhs and
```
[the remark this answers is not in the document]
Comment (esben-semmle) [reply javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: Hmm. Relying on the dataflow graph for what is a control flow graph problem seems to require some substantial changes that does not make the implementation clearer at all.

The core problem in using `DataFlow::PropWrite` everywhere is that for the syntactic `n1 = n2`, the assignment happens after the evaluation of `n2`, and this makes `n1` (which is the `DataFlow::PropWrite` node) a poor proxy for when the assignment happens, `n2` is a slightly better candidate, but additional complications are then introduced for for `n1 = n2 = n3`.

File: javascript/ql/src/Declarations/DeadStoreOfProperty.ql [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  // conservatively reject all side-effects
+  e.isImpure()
+}
+
+/**
+ * Holds if `assign1` and `assign2` both assign property `name`, but `assign1` is dead because of `assign2`.
+ */
+predicate isDeadAssignment(string name, Assignment assign1, Assignment assign2) {
+  postDominatedPropWrite(name, assign1, assign2) and
+  maybeHasPropAccessBetween(name, assign1, assign2) and
```
[the remark this answers is not in the document]
Comment (esben-semmle) [reply javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: Fixed. The name was missing a negation.

File: javascript/ql/src/Declarations/DeadStoreOfProperty.ql (hunk 1 above)
[the remark this answers is not in the document]
Comment (xiemaisi) [reply javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: Ah, that's a good point. I forgot that unlike `VarDef`s `PropWrite`s do not sit at the point in the CFG where the assignment happens. That's a shame, but also difficult to change. Could we perhaps introduce a predicate on `PropWrite`s that gets this CFG node?

[the remark this answers is not in the document]
Comment (esben-semmle) [reply javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: Done. I will redo the performance experiments.

[the remark this answers is not in the document]
Comment (esben-semmle) [reply javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: Performance is unchanged: https://git.semmle.com/esben/dist-compare-reports/tree/js/unconditional-property-override_1541555817140

File: javascript/ql/src/Declarations/DeadStore.qll [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/**
+ * Provides classes and predicates for reasoning dead stores.
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/DeadStore.qll]: This sentence is missing an "about" after "reasoning".

**Comment by esben-semmle** [comment]
Ping @Semmle/doc for a JS qhelp review.

File: javascript/ql/src/Declarations/DeadStoreOfProperty.ql [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    e = block.getNode(j) and
+    maybeAccessesProperty(e, name) |
+    after = true and i < j
+    or
+    after = false and j < i
+  )
+}
+
+/**
+ * Holds if `assign1` and `assign2` both assign property `name`, and the assigned property may be accessed between the two assignments.
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: The doc comment still says "may be accessed" where it should say "is not accessed".

File: javascript/ql/src/Declarations/DeadStoreOfProperty.ql [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+ * @tags maintainability
+ * @precision high
+ */
+
+import javascript
+import Expressions.DOMProperties
+import DeadStore
+
+/**
+ * Holds if `assign` definitely assigns property `name` of `base`.
```
Comment (xiemaisi) [inline javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: It's not very clear what "definitely" refers to in this comment. I struggle to find a clearer way of putting it, but it needs to be made clear somehow that what we're ensuring here is that the write happens on `base` (as opposed to, say, ensuring that the write refers to `name`, or that the write definitely must happen, or something like that).

Comment (esben-semmle) [reply javascript/ql/src/Declarations/DeadStoreOfProperty.ql]: The docstring is a bit more operational now.

**Pull request merged**